compiler/rustc_type_ir/src/infcx.rs

```diff
@@ -0,0 +1,40 @@
+use crate::{ConstVid, Interner, TyVid, UniverseIndex};
+
+pub trait InferCtxtLike {
+    type Interner: Interner;
+
+    fn interner(&self) -> Self::Interner;
+
+    fn universe_of_ty(&self, ty: TyVid) -> Option<UniverseIndex>;
+
+    /// Resolve `TyVid` to its root `TyVid`.
+    fn root_ty_var(&self, vid: TyVid) -> TyVid;
+
+    /// Resolve `TyVid` to its inferred type, if it has been equated with a non-infer type.
+    fn probe_ty_var(&self, vid: TyVid) -> Option<<Self::Interner as Interner>::Ty>;
+
+    fn universe_of_lt(
+        &self,
+        lt: <Self::Interner as Interner>::InferRegion,
```

Comment:
maybe do what we do for `IntoIterator` and change this trait to
```rust
trait InferCtxtLike {
    type Interner: Interner<Ty = Self::Ty, InferRegion = Self::InferRegion, ..>;
    type InferRegion;
    type Ty;
}
```
that should make it a lot easier to read code using `InferCtxtLike`, which is imo what we should optimize for.